prusti-viper/src/encoder/backtranslation.rs

```diff
@@ -0,0 +1,457 @@
+                            variant,
+                            sil_entry,
+                            vir_name,
+                            subst,
+                            silicon_ce_entries,
+                        );
+                        Entry::Struct{
+                            name: struct_name,
+                            field_names,
+                            field_entries
+                        }
+                    },
+                    AdtKind::Enum => {
+                        let variants = &adt_def.variants.iter();
+                        let super_name = format!("{:?}", adt_def);
+                        let mut variant_name = String::from("?");
+                        let mut field_names = vec![];
+                        let mut field_entries = vec![];
+                        if let Some(ModelEntry::RefEntry(name, map)) = sil_entry {
+                            let disc_string = String::from("discriminant");
+                            let mut variant = None;
+                            let mut opt_discriminant = self.backtranslate_int(map.get(&disc_string));
+                            //need to find a discriminant to do something
+                            if !opt_discriminant.is_some(){
+                                //try to find disc in the associated local variable
+                                let opt_discr_locations = self.disc_info.get(&(self.def_id, vir_name.clone()));
+                                if let Some(discr_locations) = opt_discr_locations {
+                                    for name in discr_locations {
+                                        let disc_entry = silicon_ce_entries.get(name);
+                                        let value = self.backtranslate_int(disc_entry);
+                                        if value.is_some() {
+                                            opt_discriminant = value;
+                                            break;
+                                        }
+                                    }
+                                }
+                            }
+                            if let Some(x) = opt_discriminant {
+                                let discriminant = x as u32;
+                                //is this a risk? I assume discriminant will not be too large
+                                for def in &adt_def.variants{
+                                    let discr = match def.discr {
+                                        ty::VariantDiscr::Relative(d) => d,
+                                        _ => unimplemented!()  
+                                    };
+                                    if discr == discriminant {
+                                        variant = Some(def);
+                                        variant_name = def.ident.name.to_ident_string();
+                                    }
+                                }
```

Comment:
Yes, this is not good. Is there no API to lookup variant given the discriminant?

Reply:
I remember spending quite a bit of time on this and there is no simple method for that. I will look into it again, something better than what I currently have must definitely be possible.

Reply:
I simplified it a bit, but it essentially still does the same. I really don't know how to do this another way. I am very open to suggestions.